src/accountsdb/index.zig

```diff
@@ -146,214 +170,170 @@ pub const AccountIndex = struct {
         const reference_memory, var reference_memory_lg = self.reference_memory.writeWithLock();
         defer reference_memory_lg.unlock();
 
-        if (!reference_memory.remove(slot)) {
-            return error.MemoryNotFound;
-        }
+        const removed_kv = reference_memory.fetchRemove(slot) orelse return error.MemoryNotFound;
+        removed_kv.value.deinit();
     }
 
-    pub fn getReference(self: *Self, pubkey: *const Pubkey) ?AccountReferenceHead {
+    /// Get a read-safe account reference head, and its associated lock guard.
+    /// If access to many different account reference heads which are potentially in the same bin is
+    /// required, prefer instead to use `getBinFromPubkey(pubkey).read*(){.get(pubkey)}` directly.
+    pub fn getReferenceHeadRead(self: *Self, pubkey: *const Pubkey) ?struct { AccountReferenceHead, RwMux(RefMap).RLockGuard } {
         const bin, var bin_lg = self.getBinFromPubkey(pubkey).readWithLock();
-        defer bin_lg.unlock();
-        return bin.get(pubkey.*);
+        const ref_head = bin.get(pubkey.*) orelse {
+            bin_lg.unlock();
+            return null;
+        };
+        return .{ ref_head, bin_lg };
+    }
+
+    /// Get a write-safe account reference head, and its associated lock guard.
+    /// If access to many different account reference heads which are potentially in the same bin is
+    /// required, prefer instead to use `getBinFromPubkey(pubkey).write*(){.get(pubkey)}` directly.
+    pub fn getReferenceHeadWrite(self: *Self, pubkey: *const Pubkey) ?struct { AccountReferenceHead, RwMux(RefMap).WLockGuard } {
+        const bin, const bin_lg = self.getBinFromPubkey(pubkey).writeWithLock();
+        const ref_head = bin.get(pubkey.*) orelse return null;
+        return .{ ref_head, bin_lg };
+    }
+
+    pub const GetAccountRefError = error{ SlotNotFound, PubkeyNotFound };
```

Comment:
move to top of struct

Reply:
599503fa293d0b9e5ede22c54afa525937c810f4